File: examples/safe-agent/mcp-source/read_only_tools.py
from mcp.server.fastmcp import FastMCP

mcp = FastMCP("source-python-readonly-tools")


@mcp.tool(
    name="python_read_internal_doc",
    description="Read an approved internal documentation record by identifier.",
    annotations={"readOnlyHint": True, "idempotentHint": True},
)
def read_internal_doc(document_id: str) -> str:
    return "approved internal summary"

File: examples/safe-agent/mcp-source/read-only-tools.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";

const server = new McpServer({ name: "internal-readonly-source-tools", version: "0.1.0" });

server.registerTool(
  "source_read_internal_doc",
  {
    description: "Read an approved internal documentation record by identifier.",
    inputSchema: z.object({
      document_id: z.string()
    }).strict(),
    annotations: {
      readOnlyHint: true,
      idempotentHint: true
    }
  },
  async () => ({ content: [{ type: "text", text: "approved internal summary" }] })
);
